reorganizing the documentation

dbu opened the pull request:
fix https://github.com/php-http/httplug/issues/55 and #9

Nyholm commented:
Awesome!

dbu commented on docs/tutorial.md:
```diff
@@ -0,0 +1,46 @@
+<?php
+require('vendor/autoload.php');
+
+TODO: create client instance with discovery and do some requests
```
some todo remaining. not sure if i get to work on this much tomorrow. maybe we should merge with the TODOs left in here.

sagikazarmark replied:
Sure.

dbu commented on docs/upgrading.md:
```diff
@@ -5,56 +5,16 @@ There are some major changes between the two library. This guide will help you u
 ## New vendor
 
-The new organization name is **PHP HTTP**. This suggests that while our main product is the HTTP Adapter project, we are not limited to it. We plan to collect any kind of HTTP related packages written in PHP.
+The new organization name is **PHP HTTP**. This suggests that while our main product is the Httplug project, we are not limited to it. We plan to collect any kind of HTTP related packages written in PHP.
```
this chapter could be expanded a bit, but not urgent.

sagikazarmark replied:
Such information should probably go on the website, not the documentation.

dbu replied:
i think the right place actually would be an UPGRADE.md file in the
Httplug repository. we should mention that file in the documentation
however. does that make sense to you?

sagikazarmark replied:
Don't know. I usually try to avoid including UPGRADE file in the repository, as there are no BC breaks between major releases, minor releases are usually rare and such information can go into the documentation. We should have something like this in the documentation as well, but the history and description are probably out of scope. We should mention that httplug replaces ivory, and some updating information, done.

dbu replied:
i realized that this is not about upgrading, but about migrating. i
started a migration guide that certainly could be elaborated some more.

dbu commented:
> Awesome!

thanks. i tried adding a bit of an overview. does that respond to your
question? glad for any feedaback!

Nyholm commented:
I agree, nothing is ugent. But this documentation helps a lot. 
I think we could merge it with todo. As long as they are resolved before a release.

Nyholm commented:
It is a great start. The concept of virtual packages are quite unused. You have explained it well but I believe you have to be even more explanatory. But that may be covered in the "how to implement this in a reusable package" section.

joelwurtz commented on docs/index.md:
```diff
@@ -1,87 +1,34 @@
 # HTTP Adapter
```
Change the title to Httplug or Http Client ?

sagikazarmark replied:
I would go with Httplug. Httpl\Client is used only for namespace.

sagikazarmark commented:
@Nyholm I remember back then I started explaining the rationale behind virtual package architecture, also gave some examples what to require in an application, what to require in a reusable package. Not sure if it is removed, but I think not.

joelwurtz commented on docs/index.md:
```diff
@@ -1,87 +1,34 @@
+**This is the documentation for the Httplug web client abstraction and the other PHP-HTTP components.**
 
-The HTTP Adapter abstracts from PHP HTTP clients that are based on [PSR-7](http://www.php-fig.org/psr/psr-7/).
+Httplug abstracts from HTTP clients written in PHP that are based on [PSR-7](http://www.php-fig.org/psr/psr-7/).
```
I don't think it is correct, like we want to provide guzzle 5 adapter where the client is not based on PSR-7.

sagikazarmark replied:
I guess this wants to be that the abstraction is created based on existing PSR7 clients (which I wonder also if it's true, as the first PSR7 client was guzzle, and we started to work on the interface long before PSR7 was out)

sagikazarmark commented:
@dbu :+1:

joelwurtz commented on docs/package-overview.md:
```diff
@@ -0,0 +1,44 @@
+- Repository: https://github.com/php-http/message-factory
+- Documentation: [Message Factory](message-factory.md) 
+
+## *-adapter
```
*-adapter or *-client

Or maybe we should just rename socket-client to socket-adapter (it may be not an adapter to a package, but in a sense it is an adapter for socket / stream function in PHP)

sagikazarmark replied:
I think we could use both. The basic is the client, since it is a client interface. The next level is an adapter (which is still a client).

joelwurtz commented on docs/upgrading.md:
```diff
@@ -5,56 +5,16 @@ There are some major changes between the two library. This guide will help you u
 ## Package separation
 
-One of the biggest changes changes is the package separation. One big package is separated ito several smaller one.
+One of the biggest changes changes is the package separation. One big package is separated ito several smaller ones.
```
into smaller ones ?

hannesvdvreken replied:
Also 2 x 'changes'

joelwurtz commented on docs/upgrading.md:
```diff
@@ -5,56 +5,16 @@ There are some major changes between the two library. This guide will help you u
+The packages are listed in the (package overview)[package-overview.md]
 
 
 ## Removed HTTP methods from adapter
```
Old interfaces below

joelwurtz commented on docs/utils.md:
```diff
@@ -0,0 +1,25 @@
+# PHP-HTTP Utilities
+
+The utilities contain some useful tools for working with HTTP. Include them in your project with composer:
```
Utilities / The utility package ?

dbu commented on docs/package-overview.md:
```diff
@@ -0,0 +1,44 @@
+
+**Each adapter is separated into its own package. This allows to make requirements to the underlying HTTP Client libraries.**
+
+- Namespace: `Http\Adapter`
```
is this namespace still correct? or would it be Http\Client?

dbu commented:
i added a section specifically for the virtual package topic. also with the idea that people can link to that in their projects.

updated for all the other comments. unless there is some big problem, i would love to merge this and work on the tutorial in follow-up PR and also for moving the upgrading documentation to a file in the code repository.

Nyholm commented on docs/virtual-package.md:
```diff
@@ -0,0 +1,9 @@
+
+When *writing a reusable library*, the `require` section should only mention `php-http/client-implementation`, while the `require-dev` section needs to specify a concrete implementation in most cases, in order for the package to be installable on its own for development and testing.
+
+When *using a reusable library* in an application, the `require` section needs to include a concrete implementation in order for the application to install. Missing to do that will lead to composer reporting a missing package.
```
This line feels strange. It's correct but strange.. Maybe: 

> When *using a reusable library* in your application, one must require a concrete implementation. Make sure the `require` section includes one [package that provides `php-http/client-implementation`](https://packagist.org/providers/php-http/client-implementation). Failing to do that will lead to composer reporting a missing package.

Nyholm replied:
You maybe should include the composer error message. I believe many people are going to search for "The requested package php-http/client-implementation could not be found in any version, there may be a typo in the package name."

```
$ composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package php-http/client-implementation could not be found in any version, there may be a typo in the package name.
```

sagikazarmark replied:
I agree, it is a bit weird. I would close to the problem from a library point of view: libraries does not depend on an actual implementation, but a virtual one. This is to avoid hard coupling and to allow using an implementation chosen by the user. Applications however must require one providing that virtual package. You can think about this as an "interface/contract" for packages.

I don't think including the composer error message adds any value. This is a composer detail. The actual message doesn't even matter as it leads directly to the fix, so it does not help identifying the problem, it's itself the identification. Enough knowledge to use composer is required.

dbu replied:
i think the error message is a very good idea for one reason: people will google that error message and they should find a page that explains what is going wrong. would be awesome if one day we could register virtual packages in packagist and have composer output an explanation, but until that day we should have this

sagikazarmark replied:
Okay, if something, then a gif animation, because it looks good :stuck_out_tongue_winking_eye: 

For copy pasters we can have the textual error as well.

joelwurtz commented on docs/index.md:
```diff
@@ -1,87 +1,35 @@
+Httplug allows you to write reusable libraries and applications that need a HTTP client without binding to a specific implementation. When all packages used in an application only specify Httplug, the application developers can chose the client that fits best for their project and use the same client with all packages.
 
-in 2015 a decision has been made to move the library to it's own organization, so PHP HTTP was born.
+There are clients implementing the Httplug `Http\Client\Client` interface directly, and adapter packages that implement the interface and forward the calls to HTTP clients not implementing the interface.
```
Http\Client\HttpClient

joelwurtz commented:
:+1:

sagikazarmark commented on docs/httplug.md:
```diff
@@ -0,0 +1,52 @@
+1. Usage in a project
+2. Usage in a reusable package
+
+In both cases, the client provides a `sendRequest` method to send a PSR-7 `RequestInterface` and returns a PSR-7 `ResponseInterface` or throws an exception that implements `Http\Client\Exception`. The client also has a `sendRequests` to send requests in parallel. Clients that do not support sending requests in parallel will just send them sequentially.
```
Send requests is not true anymore.

sagikazarmark commented on docs/httplug.md:
```diff
@@ -0,0 +1,52 @@
+
+## Httplug implementations
+
+Httplug implementations typically are either HTTP clients of their own, or they are adapters wrapping existing clients like Guzzle 6. In the later case, they will depend on the required client implementation, you should only need to require the adapter and not the actual client.
```
Isn't it **latter case**?

sagikazarmark replied:
> you ~~should~~ only need

WDYT?

sagikazarmark commented on docs/httplug.md:
```diff
@@ -0,0 +1,52 @@
+
+Httplug implementations typically are either HTTP clients of their own, or they are adapters wrapping existing clients like Guzzle 6. In the later case, they will depend on the required client implementation, you should only need to require the adapter and not the actual client.
+
+Clients usually also provide a (message factory)[message-factory.md] to allow you to create requests. See (packagist)[https://packagist.org/providers/php-http/client-implementation] for the full list of implementations.
```
I am a bit confused by this paragraph. At least I don't see how the two sentences in it are connected. Am I missing something?

sagikazarmark commented on docs/httplug.md:
```diff
@@ -0,0 +1,52 @@
+
+## Usage in a project
+
+When installing in an application or a non-reusable package, you should require a concrete client implementation. The client will in turn depend on `php-http/httplug`.
```
I am not sure referring to packages by putting them between backticks is the way we should do it (although I am the one who brought this into practice). Options:

1. Leave it as is
2. Link to github repo
3. Link to packagist

dbu replied:
i think backticks is not that bad. links to packagist certainly would be
elegant. we can fix that later, most important is its consistent.

sagikazarmark replied:
I agree consistency is more important. #12

sagikazarmark commented on docs/httplug.md:
```diff
@@ -0,0 +1,52 @@
+A few things should be taken into consideration before choosing an adapter:
+
+- It is possible that some other dependency already has an HTTP Client requirement like Guzzle 6. It can be confusing to have more than one HTTP Client installed, so always check your other requirements and choose an adapter based on that.
+- Some clients support parallel requests, others only emulate them. If parallel requests are needed, use one which supports it.
```
This is probably obsolete as well.

sagikazarmark commented on docs/httplug.md:
````diff
@@ -0,0 +1,52 @@
+...
+```
+
+If injecting the client instance into the code of the reusable package is not convenient, you can use the (Discovery)[discovery.md] system to free the user of your package from having to instantiate the client.
````
I would add an extra note here: "However, this way they are losing the possibility to configure it."

sagikazarmark replied:
Further thinking about this: the point is that they cannot configure the client later. Probably they are aware of the fact that by using the discovery they cannot configure it at construction time.

dbu replied:
reformulated this one, hope its more clear now.

sagikazarmark commented on docs/index.md:
```diff
@@ -1,87 +1,35 @@
+# PHP-HTTP Httplug
 
-**This is the documentation for HTTP Adapter and it's software components.**
+**This is the documentation for the Httplug web client abstraction and the other PHP-HTTP components.**
```
"web client"?

sagikazarmark commented on docs/package-overview.md:
```diff
@@ -0,0 +1,49 @@
+
+**Each client implementation is separated into its own package. This allows for clean requirement specification.**
+
+- Namespace: `Http\Client`
```
I wouldn't use this namespace. It is possible, but it probably slows composer down then multiple packages use the same namespace for autoloading. Maybe @Seldaek can help us out with this.

Seldaek replied:
Sure conflicts in autoloading if you have many packages installed at the same time that configure the same namespaces means that the autoloader will have to look into a few directories before it finds a match for a class, so that might not be good, but I have zero context as to what you are doing here so not sure if it's really a problem or not.

sagikazarmark replied:
Thanks @Seldaek 

The plan is to have adapters to different existing HTTP Clients and have them all under Http\Client namespace, but in separate package.

Seldaek replied:
OK so most likely you wouldn't have multiple installed at once so it 
shouldn't be an issue.

Not sure if using a namespace like "Http" is a great idea though, given 
this is just a common name and not really vendor specific, but up to you :)

dbu replied:
i just delete the namespace here in the doc. adapters and clients can
live anywhere they want.

sagikazarmark commented:
@dbu nice work. I added a few comments which we should discuss first, or open a separate issue.

dbu commented:
thanks a lot for the close read. i think there was a lot of clarifcation, great work! i created a bunch of new issues, propose that we merge this now.

sagikazarmark commented:
Go ahead.

dbu commented:
funny, from working with my team i am used to never click merge on a pull request i created :-)
we are very much into peer reviews for the stuff we do...

sagikazarmark commented:
At work I also follow this principle. In OSS I don't like to be that strict. I review the PR anyway, but devs might want to add something else in the project and they are free to. Also, guys working on OSS are usually sane developers, who require less control to produce sane code :stuck_out_tongue: 

So review is only for mistakes, rarely for sanity check.
